### PR TITLE
[clang-tidy] Adjust size-empty doc because C++11 size() is constant-time

### PR DESCRIPTION
**Pull request by N-Dekker**
From C++11, a conforming `size()` method is guaranteed to be a constant-time function. `empty()` is not _generally_ more efficient than `size()`. It might even be implemented in terms of `size()`.

----

Notes:
- Microsoft's STL had implemented `empty()` as `return size() == 0`, until May 2021: https://github.com/microsoft/STL/pull/1836/files
- The time complexity of `size()` (specifically for `std::set`) was discussed by the library working group in 2007-2009: https://cplusplus.github.io/LWG/issue632

### CODE REVIEW

**Review by dzenanz — APPROVED** [review]

**Comment by N-Dekker** [comment]
> You understand that this check does not apply only to std:: containers but also to boost and other custom one that have size and empty methods. In such case claim is still valid.

Thanks for your prompt reply, Piotr. I thought of custom containers too, but then again, I would rather not make any claim about the efficiency of `size()` and `empty()` methods of _arbitrary_ custom containers. I mean, they could do anything!

In order for the documentation to make claims about  `size()` versus `empty()`, I think they should behave like those of std:: containers, including their time complexity. Otherwise it's hard to reason about them in a general way. 

But then, do you happen to know whether there are indeed still Boost containers that have a `size()` method which has a greater-than-constant time complexity? If so, can you possibly name one?

> If you change check description, then check documentation also should be updated to be in sync.

Thanks! So this pull request should make some more adjustments before it can be merged (if it would be accepted at all)?

**Review by HerrCai0907 — COMMENTED** [review]
Please update `clang-tools-extra/docs/clang-tidy/checks/readability/container-size-empty.rst` also.

**Comment by HerrCai0907** [comment]
> In such case claim is still valid

I think we should remove the discussion about efficient in claim since it is misleading after C++11. For self-defined container, because implement of empty and size can be totally different, this claim is still misleading.

**Comment by N-Dekker** [comment]
Thanks for your encouragement, @HerrCai0907. I just [force-pushed](https://github.com/llvm/llvm-project/compare/87688f9ec5e74f9b3c74df05a69a4f822966b5f1..ab3009e825af5e323e749dfdbd2300ef68677f14), as you suggested.

To further clarify my motivation, I believe that C++ users actually got the impression from the documentation that this check will make their code faster. ITK pull request https://github.com/InsightSoftwareConsortium/ITK/pull/4985 ("PERF: readability container size empty") was also labeled as a performance improvement, while I think it is very unlikely to have any effect on the run-time performance.

**Review by PiotrZSL — APPROVED** [review]
LGTM

**Review by HerrCai0907 — APPROVED** [review]
LGTM. Thanks!

**Comment by N-Dekker** [comment]
@PiotrZSL @HerrCai0907 Thank you both for your feedback and approval! Would one of you be able the merge the pull request?

**Pull request merged**
